onnxruntime/core/providers/cpu/tensor/cast_op.cc

```diff
@@ -116,115 +148,117 @@ CastFromString(const std::string& input, DstType& output) {
   output = gsl::narrow_cast<DstType>(std::stoll(input));
 }
 
-// generic scalar X -> Y
-template <typename SrcType, typename DstType>
-struct ScalarDirectCaster {
-  void Cast(const SrcType& in, DstType& out) const {
-    out = static_cast<DstType>(in);
-  }
-};
-
-// scalar X -> string
-template <typename SrcType>
-struct ScalarDirectCaster<SrcType, std::string> {
-  void Cast(const SrcType& in, std::string& out) const {
-    CastToString<SrcType>(in, out);
-  }
-};
+// overloads for MLFloat16 and BFloat16
+void CastFromString(const std::string& input, MLFloat16& output) {
+  float intermediate;
+  CastFromString(input, intermediate);
+  output = static_cast<MLFloat16>(intermediate);
+}
 
-// scalar string -> X
-template <typename DstType>
-struct ScalarDirectCaster<std::string, DstType> {
-  void Cast(const std::string& in, DstType& out) const {
-    CastFromString<DstType>(in, out);
-  }
-};
+void CastFromString(const std::string& input, BFloat16& output) {
+  float intermediate;
+  CastFromString(input, intermediate);
+  output = static_cast<BFloat16>(intermediate);
+}
 
-// helper for indirect cast types
-template <typename SrcType, typename DstType, typename IntermediateType>
-struct ScalarIndirectCaster {
-  void Cast(const SrcType& in, DstType& out) const {
-    IntermediateType intermediate;
-    ScalarDirectCaster<SrcType, IntermediateType>{}.Cast(in, intermediate);
-    ScalarDirectCaster<IntermediateType, DstType>{}.Cast(intermediate, out);
-  }
+template <typename T>
+struct EigenType {
+  using type = T;
 };
 
-template <typename SrcType, typename DstType, class Enable = void>
-struct ScalarCaster;
-
-template <typename SrcType, typename DstType>
-struct ScalarCaster<
-    SrcType, DstType,
-    typename std::enable_if<AreAllDirectCastTypes<SrcType, DstType>::value>::type> {
-  void Cast(const SrcType& in, DstType& out) const {
-    ScalarDirectCaster<SrcType, DstType>{}.Cast(in, out);
-  }
+template <>
+struct EigenType<MLFloat16> {
+  using type = Eigen::half;
 };
 
-template <typename SrcType, typename DstType>
-struct ScalarCaster<
-    SrcType, DstType,
-    typename std::enable_if<!AreAllDirectCastTypes<SrcType, DstType>::value>::type> {
-  void Cast(const SrcType& in, DstType& out) const {
-    ScalarIndirectCaster<SrcType, DstType, float>{}.Cast(in, out);
-  }
+template <>
+struct EigenType<BFloat16> {
+  using type = Eigen::bfloat16;
 };
 
 // generic tensor X -> Y
-template <typename SrcType, typename DstType>
+template <typename SrcType, typename DstType, typename Enable = void>
 struct TensorCaster {
-  void Cast(const Tensor& in, Tensor& out, const TensorShape& shape) const {
+  void Cast(const OpKernelContext&, const Tensor& in, Tensor& out, const TensorShape& shape) const {
+    using EigenSrcType = typename EigenType<SrcType>::type;
+    using EigenDstType = typename EigenType<DstType>::type;
+
```

Comment:
comments explaining why we use EigenType with a reinterpret_cast would be good.

Reply:
would also be good to have comments above explaining why there's a specialization for EigenType<MLFloat16> and EigenType<BFloat16>

---
In reply to: [569865762](https://github.com/microsoft/onnxruntime/pull/6509#discussion_r569865762) [](ancestors = 569865762)

Reply:
renamed stuff and added comments, hope it's clearer now

---
In reply to: [569866987](https://github.com/microsoft/onnxruntime/pull/6509#discussion_r569866987) [](ancestors = 569866987,569865762)